lib/msf/core/exploit/remote/smb/client/kerberos_authentication.rb

```diff
@@ -0,0 +1,141 @@
+    response = smb1_session_setup_response(raw_kerberos_response)
+    @kerberos_authenticator.validate_response!(response.data_block.security_blob)
+
+    response_code = response.status_code
+
+    # Store the available OS information before going forward.
+    @peer_native_os = response.data_block.native_os.to_s
+    @peer_native_lm = response.data_block.native_lan_man.to_s
+
+    @user_id = response.smb_header.uid if response_code == WindowsError::NTStatus::STATUS_SUCCESS
+
+    response_code
+  end
+
+  #
+  # @param type3_message [String] the NTLM Type 3 message
+  # @param user_id [Integer] the temporary user ID from the Type 2 response
+  # @return [String] the raw binary response from the server
+  def smb1_kerberos_authenticate(security_buffer)
+    packet = smb1_kerberos_authenticate_packet(security_buffer)
+    send_recv(packet)
+  end
+
+  # Generates the {RubySMB::SMB1::Packet::SessionSetupRequest} packet
+  # with the NTLM Type 3 (Auth) message in the security_blob field.
+  #
+  # @param type3_message [String] the NTLM Type 3 message
+  # @param user_id [Integer] the temporary user ID from the Type 2 response
+  # @return [RubySMB::SMB1::Packet::SessionSetupRequest] the second authentication packet to send
+  def smb1_kerberos_authenticate_packet(security_blob)
+    packet = RubySMB::SMB1::Packet::SessionSetupRequest.new
+    # packet.smb_header.uid = user_id
+    packet.set_security_buffer(security_blob)
+    packet.parameter_block.max_buffer_size = self.max_buffer_size
+    packet.parameter_block.max_mpx_count = 50
+    packet.smb_header.flags2.extended_security = 1
+    packet
+  end
+
+  #
+  # SMB 2 Methods
+  #
+
+  # Handles SMB2 Kerberos Authentication by delegating to a kerberos_authenticator implementation
+  # to generate a GSS security blob with an embedded AP_REQ. On success information is stored
+  # about the peer/server.
+  def smb2_authenticate
+    raise ::RubySMB::Error::AuthenticationFailure, 'Missing kerberos authenticator' unless @kerberos_authenticator
+
+    kerberos_result = @kerberos_authenticator.authenticate
+
+    raw_kerberos_response = smb2_kerberos_authenticate(kerberos_result[:security_blob])
+    response = smb2_session_setup_response(raw_kerberos_response)
+    @kerberos_authenticator.validate_response!(response.buffer)
+
+    @session_id = response.smb2_header.session_id
+    @session_key = kerberos_result[:session_key].value[0..16]
+
+    @session_is_guest = response.session_flags.guest == 1
+
+    if @smb3
+      if response.session_flags.encrypt_data == 1
+        # if the server indicates that encryption is required, enable it
+        @session_encrypt_data = true
+      elsif (@session_is_guest && password != '') || (username == '' && password == '')
+        # disable encryption when necessary
+        @session_encrypt_data = false
+      end
+      # otherwise, leave encryption to the default value that it was initialized to
+    end
```

Comment:
I recently added in the application key calculation in RubySMB [here](https://github.com/rapid7/ruby_smb/blob/5feba411750ca08755e7fedbc8ebd266d438b2b7/lib/ruby_smb/client/authentication.rb#L213) and [here](https://github.com/rapid7/ruby_smb/blob/5feba411750ca08755e7fedbc8ebd266d438b2b7/lib/ruby_smb/client/authentication.rb#L233-L243). It's needed for some DCERPC operations (right now the SAMR set-password functionality). 

That code should also probably be duplicated here so the key is available for use.

Reply:
Aligned, thanks :+1:

Will raise a ticket to split out the authentication mechanism from the post-amble here